core/vulkan/vk_virtual_swapchain/cc/base_swapchain.cpp

```diff
@@ -0,0 +1,287 @@
+      0, // bufferMemoryBarrierCount
+      nullptr, // pBufferMemoryBarriers
+      1, // imageMemoryBarrierCount
+      &initialBarrier // pImageMemoryBarriers
+  );
+
+  VkImageSubresourceLayers subresource = {
+    VK_IMAGE_ASPECT_COLOR_BIT, // aspectMask
+    0, // mipLevel
+    0, // baseArrayLayer
+    swapchain_info_.imageArrayLayers, // layerCount
+  };
+  VkOffset3D offsets[2] = {
+    { 0, 0, 0, },
+    { (int32_t)swapchain_info_.imageExtent.width,
+      (int32_t)swapchain_info_.imageExtent.height,
+      1,
+    },
+  };
+  VkImageBlit blit = {
+    subresource,
+    { offsets[0], offsets[1] },
+    subresource,
+    { offsets[0], offsets[1] },
+  };
+  device_functions_->vkCmdBlitImage(
+      command_buffer_,
+      image, // srcImage
+      VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, // srcImageLayout
+      images_[index], // dstImage
+      VK_IMAGE_LAYOUT_PRESENT_SRC_KHR, // dstImageLayout
+      1, // regionCount
+      &blit, // pRegions
+      VK_FILTER_NEAREST // filter
+  );
+
+  VkImageMemoryBarrier finalBarrier = initialBarrier;
+  finalBarrier.srcAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT;
+  finalBarrier.dstAccessMask = VK_ACCESS_MEMORY_READ_BIT;
+  finalBarrier.oldLayout = VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL;
+  finalBarrier.newLayout = VK_IMAGE_LAYOUT_PRESENT_SRC_KHR;
+
+  device_functions_->vkCmdPipelineBarrier(
+      command_buffer_,
+      VK_PIPELINE_STAGE_TRANSFER_BIT, // srcStageMask
+      0, // dstStageMask
+      0, // dependencyFlags
+      0, // memoryBarrierCount
+      nullptr, // pMemoryBarriers
+      0, // bufferMemoryBarrierCount
+      nullptr, // pBufferMemoryBarriers
+      1, // imageMemoryBarrierCount
+      &finalBarrier // pImageMemoryBarriers
+  );
+
+  if ((res = device_functions_->vkEndCommandBuffer(command_buffer_)) !=
+      VK_SUCCESS) {
+    return res;
+  }
+
+  VkPipelineStageFlags waitStage = VK_PIPELINE_STAGE_TRANSFER_BIT;
+  VkSubmitInfo submitInfo = {
+    VK_STRUCTURE_TYPE_SUBMIT_INFO, // sType
+    0, // pNext
+    1, // waitSemaphoreCount
+    &acquire_semaphore_, // pWaitSemaphores
+    &waitStage, // pWaitDstStageMask
+    1, // commandBufferCount
+    &command_buffer_, // pCommandBuffers
+    1, // signalSemaphoreCount
+    &blit_semaphore_, // pSignalSemaphores
+  };
+  auto queue_functions = GetGlobalContext().GetQueueData(queue);
+  if ((res = queue_functions->vkQueueSubmit(queue, 1, &submitInfo,
+                                            blit_fence_)) != VK_SUCCESS) {
+    return res;
+  }
+
+  VkPresentInfoKHR presentInfo = {
+    VK_STRUCTURE_TYPE_PRESENT_INFO_KHR, // sType
+    0, // pNext
+    1, // waitSemaphoreCount
+    &blit_semaphore_, // waitSemaphores
+    1, // swapchainCount
+    &swapchain_, // pSwapchains,
+    &index, // pImageIndices
+    nullptr, // pResults
+  };
+  if ((res = queue_functions->vkQueuePresentKHR(queue, &presentInfo)) !=
+      VK_SUCCESS) {
+    return res;
+  }
+
+  device_functions_->vkWaitForFences(device_, 1, &blit_fence_, VK_TRUE,
```

Comment:
I am a bit leery about this vkWaitForFences. Since this is running on the main thread, we can end up blocking the user thread and synchronizing with the GPU.